File: LeetCode/341. 扁平化嵌套列表迭代器.rs
use std::collections::VecDeque;
// #[derive(Debug, PartialEq, Eq)]
// pub enum NestedInteger {
//   Int(i32),
//   List(Vec<NestedInteger>)
// }
struct NestedIterator {
    list: VecDeque<NestedInteger>
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NestedIterator {

    fn new(nestedList: Vec<NestedInteger>) -> Self {
        Self{
            list: VecDeque::from(nestedList)
        }
    }
    
    fn next(&mut self) -> i32 {
        match self.list.pop_front().unwrap() {
            NestedInteger::Int(n) => n,
            _ => panic!("never!")
        }
    }
    
    fn has_next(&mut self) -> bool {
        while !self.list.is_empty() {
            if let NestedInteger::Int(num) = self.list[0] {
                return true;
            }
            if let NestedInteger::List(mut first) = self.list.pop_front().unwrap() {
                while !first.is_empty() {
                    self.list.push_front(first.pop().unwrap());
                }
            }
        }
        !self.list.is_empty()
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * let obj = NestedIterator::new(nestedList);
 * let ret_1: i32 = obj.next();
 * let ret_2: bool = obj.has_next();
 */

File: LeetCode/341. 扁平化嵌套列表迭代器.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """
from collections import deque
class NestedIterator:
    list = deque([])
    def __init__(self, nestedList: [NestedInteger]):
        self.list = deque(nestedList)
    
    def next(self) -> int:
        return self.list.popleft().getInteger()
    
    def hasNext(self) -> bool:
        while len(self.list) != 0 and not self.list[0].isInteger():
            first = self.list.popleft().getList()
            for x in reversed(first):
                self.list.appendleft(x)
        return len(self.list) != 0

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())
